### ISSUE TITLE
Suggestions

### ISSUE THREAD

**Issue by Z0lid** [davidepianca98/KMQTT#11]
Make `internal fun getSession(clientId: String?): ISession?` public,
and it would be nice to have a `onDisconnect` event that would trigger whenever a client has disconnected or timed out.

**Comment by davidepianca98** [comment]
Hello, thank you for the suggestions, may I know why would you want to access the sessions, in particular what would you want to do with it? It can't just be made public, because otherwise one could change potential settings while the broker is running, so I would need to expose only certain methods, mainly read only.
The  `onDisconnect` event handler would indeed be useful, so that I will definitely implement

**Comment by Z0lid** [comment]
> one could change potential settings while the broker is running

Yeah I agree with that
> may I know why would you want to access the sessions

I needed a way of checking if a client is connected, so am suggesting something like this:
```
fun isConnected(clientId: String): Boolean {
    return sessions[clientId]?.connected ?: false
}
```
in the Broker class.
Also, is there any reason for the `properties: MQTT5Properties` to not be nullable in the `publish` function?

**Comment by davidepianca98** [comment]
Sorry for the late reply, I have now implemented the requested changes, also properties in the publish function is now nullable. Do you have any other suggestion before I publish the new version?

**Comment by Z0lid** [comment]
Not as of right now, thanks for implementing the changes, i do have a question though.
Why make interfaces for stuff like `Authorization`, `Authentication` and `PacketInterceptor` when they only have a single function in them, for example when i had to quickly implement the onDisconnect i did it like this in `Broker.kt`:
`val onDisconnect: ((clientId: String, reasonCode: ReasonCode?) -> Unit)? = null`

Also, any ETA on when WebSocket support would come? (since right now i have KMQTT and WebSockets form KTor working together)

**Comment by davidepianca98** [comment]
Well, the idea is that there is only one method as of now, but one day it might be useful to add others. Having the public interface exposed as an interface it is possible without breaking existing implementations, while if it was just a single parameter to the `Broker` class I would have to add many new parameters and there are already way too many for my liking anyway. If you know a better way to do that, I'm always open to suggestions.

About WebSocket I never started working on it because there is no WS server library for Kotlin Multiplatform, so it needs quite a bit of work to get it working especially on Kotlin/Native. Also because I never needed it and nobody asked for it. Now that I know somebody would want it I can implement it for the platform you need. For example if you are using KMQTT on the JVM I could implement it there first.

**Comment by Z0lid** [comment]
Ok well yeah thats a valid argument ig, imo its just cleaner to provide the function instead of extending an interface.

Oh, alright, yeah i would be happy to have that all on one library, i need the MQTT part of it because the Wemos D1 devices on my lan dont have a stable connection over WS, but i need the WS to pass it thru cloudflare for mobile phones. I also have a HTTP -> MQTT api for stuff on the phone that doesnt need updating constantly.

**Comment by davidepianca98** [comment]
I just released version 0.3.0 with the enhancements and basic WebSocket support. Feel free to open another issue if you find anything wrong.
